src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,295 @@
+                    case ConstantPattern n: return $"C:{n.Expression}";
+                    case PositionalPattern n: return $"P:({string.Join(", ", n.Subpatterns.Select(p => p.Pattern))})";
+                    default: return null;
+                }
+            }
+#endif
+        }
+
+        private sealed class Conjunction : AnalyzedNode
+        {
+            public readonly AnalyzedNode Left;
+            public readonly AnalyzedNode Right;
+
+            public override NodeKind Kind => NodeKind.Conjunction;
+
+            public static Conjunction Create(AnalyzedNode left, AnalyzedNode right)
+            {
+                if (left is null || right is null)
+                    return null;
+
+                return new Conjunction(left, right);
+            }
+
+            public Conjunction(AnalyzedNode left, AnalyzedNode right)
+            {
+                Debug.Assert(left != null);
+                Debug.Assert(right != null);
+                Left = left;
+                Right = right;
+            }
+
+            public override bool Contains(ExpressionSyntax e) => Left.Contains(e) || Right.Contains(e);
+
+            public override AnalyzedNode Reduce(Reducer reducer) => reducer.ReduceConjunction(this);
+
+            public override void GetChildren(List<AnalyzedNode> nodes)
+            {
+                Left.GetChildren(nodes);
+                Right.GetChildren(nodes);
+            }
+
+            public override PatternSyntax AsPatternSyntax() => AsPatternSyntax(out _);
+
+            private PatternSyntax AsPatternSyntax(out List<AnalyzedNode> children)
+            {
+                children = new List<AnalyzedNode>();
+                GetChildren(children);
+                return RecursivePattern(
+                    children.OfType<TypePattern>().SingleOrDefault()?.Type,
+                    children.OfType<PositionalPattern>().SingleOrDefault()?.AsPositionalPatternClauseSyntax(),
+                    PropertyPatternClause(SeparatedList(
+                        children.OfType<PatternMatch>().Select(match => match.AsSubpatternSyntax()))),
+                    children.OfType<VarPattern>().SingleOrDefault()?.AsVariableDesignationSyntax());
+            }
+
+            public override ExpressionSyntax AsExpressionSyntax()
+                => BinaryExpression(SyntaxKind.LogicalAndExpression,
+                    Left.AsExpressionSyntax(), Right.AsExpressionSyntax());
+
+            public SyntaxNode AsCasePatternSwitchLabelSyntax()
+                => CasePatternSwitchLabel(AsPatternSyntax(out var children),
+                    AsWhenClause(children.OfType<Evaluation>().Select(e => e.Expression).AggregateOrDefault(
+                        (left, right) => BinaryExpression(SyntaxKind.LogicalAndExpression, left, right))),
+                    Token(SyntaxKind.ColonToken));
+
+            private static WhenClauseSyntax AsWhenClause(ExpressionSyntax e) => e is null ? null : WhenClause(e);
```

Comment:
docs for all of these methods.  Basically, i want to deeply feel like this is correct, and showing what we are expecting to get in and what we're translating to is important.